# (PART) Workflows {-}

# Setup

As describe in [a previous Chapter](#resourcer) the `resourcer` R package allows to deal with the main data sources (using tidyverse, DBI, dplyr, sparklyr, MongoDB, AWS S3, SSH etc.) and is easily extensible to new ones including specific data infrastructure in R or Bioconductor. So far `ExpressionSet` and `RangedSummarizedExperiment` objects saved in `.rdata` files are accesible through the `resourcer` package. The `dsOmics` package contains a new extension that deals with VCF (Variant Calling Format) files which are coerced to a GDS (Genomic Data Storage) format (VCF2GDS). 

In order to achive this `resourcer` extension, two `R6` classes have been implemented:

* `GDSFileResourceResolver` class which handles file-base resources with data in GDS or VCF formats. This class is responsible for creating a `GDSFileResourceClient` object instance from an assigned resource.
* `GDSFileResourceClient` class which is responsible for getting the referenced file and making a connection (created by `GWASTools`) to the GDS file (will also convert the VCF file to a GDS file on the fly, using `SNPRelate`). For the subsequent analysis, it's this connection handle to the GDS file that will be used.



## Required DataSHIELD packages in the opal server

Required DataSHIELD packages must be uploaded in the opal server through the Administration site by accessing to DataSHIELD tab. In our case, both `dsBase` and `dsOmics` and `resourcer` packages must be installed as is illustrated in the figure. 

```{r installPackagesOpal, echo=FALSE, fig.cap="Installed packages in the test opal server", fig.align='center'}
knitr::include_graphics("fig/add_packages_opal.png")
```


The tab **+Add package** can be used to install a new package. The figure depicts how `dsOmics` was intalled into the opal server


```{r installPackagesOpal2, echo=FALSE, fig.cap="Description how `dsOmics` package was intalled into the test opal server", fig.align='center'}
knitr::include_graphics("fig/add_packages_opal_2.png")
```

## Required R Packages in the client site (e.g. local machine)

In order to reproduce the results obtained in this part, the following R packages must be installed (in their development version):

```{r install_all, eval=FALSE}
devtools::install_github("datashield/DSI", dependencies = TRUE)
devtools::install_github("obiba/opalr", dependencies = TRUE)
devtools::install_github("datashield/DSOpal", dependencies = TRUE)
devtools::install_github("datashield/dsBaseClient", ref = "v6.0-dev", dependencies = TRUE)
devtools::install_github("isglobal-brge/dsOmicsClient", dependencies = TRUE)
```

The package dependencies are then loaded as follows:

```{r requiredRPackages}
library(DSI)
library(DSOpal)
library(dsBaseClient)
library(dsOmicsClient)
```

**Notes**:

For advanced users willing to use `DSLite`, the server side packages needs to be installed as well:

```{r install_resourcer, eval=FALSE}
install.packages("resourcer", dependencies = TRUE)
devtools::install_github("datashield/DSLite", dependencies = TRUE)
devtools::install_github("isglobal-brge/dsOmics", dependencies = TRUE)
devtools::install_github("datashield/dsBase", ref = "v6.0-dev", dependencies = TRUE)
```



## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
```
