--- 
title: "Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
output_dir: "docs"
bibliography: [resource.bib, packages.bib]
biblio-style: apalike
link-citations: yes
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(comment="", warning=FALSE, message=FALSE, cache=TRUE)
```


# Welcome {-}


<img src="fig/datashield.jpg" width="200" align="right" style="margin: 0 1em 0 1em" /></a> 

This is the website for a book that provides users with some common workflows for the non-disclosive analysis of biomedical data with [R](https://cran.r-project.org/) and [DataSHIELD](http://www.datashield.ac.uk/) from different resources. This book will teach you how to use the `r BiocStyle::CRANpkg("resourcer")` R package to perform any statistcal analysis from different studies having data in different formats (e.g., CSV, SPSS, R class, ...). In particular, we focus on illustrating how to deal with Big Data by providing several examples from omic and geographical settings. To this end, we use cutting-edge Bioconductor tools to perform transcriptomic, epigenomic and genomic data analyses. Serveral R packages are used to perform analysis of geospatial data. We also provide examples of how performing non-disclosive analyses using secure SHELL commands by allowing the use of specific software that properly deals with Big Data outside R. This material serves as an online companion for the manuscript “Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD”.


While we focus here in genomic and geoespatial data, dozens of data analysis aplications interested in performing non-disclosive analysis having data in any specific format could be carried out. By learning the grammar of DataSHILED workflows, we hope to provide you a starting point for the exploration of your own data, whether it be omic, geospatial or otherwise.

This book is organized into five parts. In the Preamble, we introduce the book, provides a tutorial for key data infrastructure useful for omic and geospatial data and a general overview for learning how Opal and DataSHILED allows performing non-disclosive data analyses from multiple studies simultaneously. So far, DataSHIELD uses tables from repository data in Opal which have some limitations to perfom Big Data analyses. 

The second part, Focus Topic, dive into information for for non-disclosive analyses using any type of resource which is one of the key advances provided in this work. It allow, among others, perform big data analyses using genomic or geospatial information where thousand of sensitive data have to be managed.  

The third part, Resources Extensions, provides examples illustrating how to extend existing resources. We describe how to create functions to deal with omic data in [Variant Calling Format](https://en.wikipedia.org/wiki/Variant_Call_Format) (VCF files) which specifies the format of a text file used in bioinformatics for storing [gene sequence](https://en.wikipedia.org/wiki/Gene) variations. It also shows how to perform genomic data analysis using secure shell commands. This can be consider as an illustrative example for using shell programs to perform Big Data analyses that could be extend to other frameworks of Big Data such as “Apache Spark”. 

The fourth part, Workflows, provides primarily code detailing the analysis of various datasets throughout the book. Initially, we provide examples from transcriptomic, epigenomic and genomic association studies, as well as examples linking geospatial data where data confidenciallity is an important issue. 


The fifth part, Developers, provides information about how to develop DataSHIELD packages using specific R packages devoted to Big Data analyses as well as using shell programs. It also provides some useful trips and trick that developers or end users may face when using our proposed infrastructure.

Finally, the Appendix highlights our contributors.

If you would like to cite this work, please use the reference “Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD”.

```{r include=FALSE}
## automatically create a bib database for R packages
knitr::write_bib(c(
  .packages(), 'bookdown', 'knitr', 'rmarkdown'
), 'packages.bib')
library(BiocStyle)
```



<!--chapter:end:index.Rmd-->

# (PART) Preamble {-}

# Introduction

# R

# Bioconductor data infrastructures {#BioC}



<!--chapter:end:00-Introduction.Rmd-->


# OPAL

Placeholder



<!--chapter:end:01-OPAL.Rmd-->


# DataSHIELD {#DataSHIELD}

Placeholder


## R and DataSHIELD implementation in Opal
## R module
## DataSHIELD module
## Web Services
## R Server Component
## R Clients (Analysis Computers)

<!--chapter:end:02-DataSHIELD.Rmd-->


# DSI: DataSHIELD Interface implementation for Opal data repository

Placeholder


## Class Structure
## Higher Level Functions
## Options

<!--chapter:end:03-DSI.Rmd-->


# (PART) Focus Topics {-}
# The resources {#resources}

Placeholder


## Concept
## Types of resources

<!--chapter:end:04-resources.Rmd-->


# The resourcer package {#resourcer}

Placeholder


## File Resources
## File Getter
## File Data Format
## Database Resources
### DBI Connectors
### Use dplyr
### Document Databases
## Computation Resources
## Extending Resources {#extending_resources}
## API declaration of a new resource
## Resource Forms
## Examples with different resources
### TSV file into a tibble or data.frame
### .Rdata file into a specific R object

<!--chapter:end:05-resourcer_package.Rmd-->

# (PART) Resources Extensions {-}

# VCF files to GDS to peform GWAS with Bioconductor {#omic_extension}

Genomic data can be stored in different formats. [PLINK](http://zzz.bwh.harvard.edu/plink/) and [VCF](https://www.internationalgenome.org/wiki/Analysis/vcf4.0/) files are commonly used in genetic epidemiology studies. In order to deal with this type of data, we have extended the resources available at the `r Githubpkg("obiba/resourcer")` package to VCF files. **NOTE**: PLINK files can be translated into VCF files using different pipelines. In R you can use `r Biocpkg("SeqArray")` to get VCF files. 

We use the [Genomic Data Storage](https://bioconductor.org/packages/release/bioc/vignettes/gdsfmt/inst/doc/gdsfmt.html#introduction) (GDS) format which efficiently manage VCF files into the R environment. This extension requires to create a [Client and a Resolver](#extending_resources) function for the `r BiocStyle::CRANpkg("resourcer")` that are located into the `r Biocpkg("dsOmics")` package. The client function uses `snpgdsVCF2GDS` function implemented in `r Biocpkg("SNPrelate")` to coerce the VCF file to a GDS object. Then the GDS object is loaded into R as an object of class `GdsGenotypeReader` from `r Biocpkg("GWASTools")` package that facilitates downstream analyses.

The opal API server allows to incorporate this new type of resource as illustrated in Figure \@ref(fig:resourceVCF).


```{r resourceVCF, echo=FALSE, fig.cap="Description of how a VCF file can be added to the opal resources", out.height= '5%', fig.align='center'}
knitr::include_graphics("fig/opal_resource_VCF.png")
```


It is important to notice that the URL should contain the tag `method=biallelic.only&snpfirstdim=TRUE` since these are required parameters of `snpgdsVCF2GDS` function. This is an example:

```
https://raw.githubusercontent.com/isglobal-brge/scoreInvHap/master/inst/extdata/example.vcf?method=biallelic.only&snpfirstdim=TRUE
```

In that case we indicate that only biallelic SNPs are considered ('method=biallelic.only') and that genotypes are stored in the individual-major mode, (i.e., list all SNPs for the first individual, and then list all SNPs for the second individual, etc) ('snpfirstdim=TRUE').

 

<!--chapter:end:06-extension_resources1_genomic.Rmd-->


# Secure shell programs: GWAS with PLINK {#shell_extension}

Placeholder



<!--chapter:end:06-extension_resources2-shell_command.Rmd-->


# (PART) Workflows {-}
# Setup

Placeholder


## Providing DataSHIELD packages in the opal server
## Required R Packages in the client site (e.g. local machine)

<!--chapter:end:07-workflow1-setup.Rmd-->


# Basic statistical analyses

Placeholder



<!--chapter:end:07-workflow2_simple_analyses.Rmd-->


# Omic data analysis {#Omic}

Placeholder


## Types of analyses implemented
## Differential gene expression (DGE) analysis
## Epigenome-wide association analysis (EWAS) 
### Single CpG analysis
### Multiple CpG analysis
### Adjusting for Surrogate Variables
## GWAS with Bioconductor
## GWAS with PLINK

<!--chapter:end:07-workflow3_omic_data_analysis.Rmd-->

# Geographical data analysis {#GIS}




<!--chapter:end:07-workflow4_GIS.Rmd-->


# (PART) Developers {-}
# DSLite: DataSHIELD Implementation on Local Datasets

Placeholder


## Development Environment Setup
### DataSHIELD Packages
### Test Datasets
### R Package Development Tools
## DataSHIELD Development Flow
### DSLiteServer
### DataSHIELD Configuration
## DataSHIELD Sessions
## Debugging
## Limitations
### Function Parameters Parser
### Server Side Environments

<!--chapter:end:11-DSLite.Rmd-->

# Creating DataSHIELD packages

<!--chapter:end:12-Creating_DS_packages.Rmd-->

# Tips and tricks

## How to install R packages into OPAL server from R

## How to check whether there are open R sesions in the OPAL server

## How 

<!--chapter:end:13-Useful_information.Rmd-->

# (PART) Appendix {-}

<!--chapter:end:14-Appendix.Rmd-->

# Session Info

```{r sessionInfo}
sessionInfo()
```

<!--chapter:end:15-SessionInfo.Rmd-->

# Bibliography

<!--chapter:end:16-References.Rmd-->

