<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>12 Basic statistical analyses | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD</title>
  <meta name="description" content="12 Basic statistical analyses | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="12 Basic statistical analyses | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="12 Basic statistical analyses | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-05-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="setup.html"/>
<link rel="next" href="Omic.html"/>
<script src="libs/header-attrs-2.1/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Wellcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="r.html"><a href="r.html"><i class="fa fa-check"></i><b>2</b> R</a></li>
<li class="chapter" data-level="3" data-path="BioC.html"><a href="BioC.html"><i class="fa fa-check"></i><b>3</b> Bioconductor data infrastructures</a></li>
<li class="chapter" data-level="4" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>4</b> OPAL</a></li>
<li class="chapter" data-level="5" data-path="DataSHIELD.html"><a href="DataSHIELD.html"><i class="fa fa-check"></i><b>5</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="5.1" data-path="DataSHIELD.html"><a href="DataSHIELD.html#r-and-datashield-implementation-in-opal"><i class="fa fa-check"></i><b>5.1</b> R and DataSHIELD implementation in Opal</a></li>
<li class="chapter" data-level="5.2" data-path="DataSHIELD.html"><a href="DataSHIELD.html#r-module"><i class="fa fa-check"></i><b>5.2</b> R module</a></li>
<li class="chapter" data-level="5.3" data-path="DataSHIELD.html"><a href="DataSHIELD.html#datashield-module"><i class="fa fa-check"></i><b>5.3</b> DataSHIELD module</a></li>
<li class="chapter" data-level="5.4" data-path="DataSHIELD.html"><a href="DataSHIELD.html#web-services"><i class="fa fa-check"></i><b>5.4</b> Web Services</a></li>
<li class="chapter" data-level="5.5" data-path="DataSHIELD.html"><a href="DataSHIELD.html#r-server-component"><i class="fa fa-check"></i><b>5.5</b> R Server Component</a></li>
<li class="chapter" data-level="5.6" data-path="DataSHIELD.html"><a href="DataSHIELD.html#r-clients-analysis-computers"><i class="fa fa-check"></i><b>5.6</b> R Clients (Analysis Computers)</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html"><i class="fa fa-check"></i><b>6</b> DSI: DataSHIELD Interface implementation for Opal data repository</a>
<ul>
<li class="chapter" data-level="6.1" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#class-structure"><i class="fa fa-check"></i><b>6.1</b> Class Structure</a></li>
<li class="chapter" data-level="6.2" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#higher-level-functions"><i class="fa fa-check"></i><b>6.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="6.3" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#options"><i class="fa fa-check"></i><b>6.3</b> Options</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="7" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>7</b> The resources</a>
<ul>
<li class="chapter" data-level="7.1" data-path="resources.html"><a href="resources.html#concept"><i class="fa fa-check"></i><b>7.1</b> Concept</a></li>
<li class="chapter" data-level="7.2" data-path="resources.html"><a href="resources.html#types-of-resources"><i class="fa fa-check"></i><b>7.2</b> Types of resources</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>8</b> The resourcer package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>8.1</b> File Resources</a></li>
<li class="chapter" data-level="8.2" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>8.2</b> File Getter</a></li>
<li class="chapter" data-level="8.3" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>8.3</b> File Data Format</a></li>
<li class="chapter" data-level="8.4" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>8.4</b> Database Resources</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>8.4.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="8.4.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>8.4.2</b> Use dplyr</a></li>
<li class="chapter" data-level="8.4.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>8.4.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>8.5</b> Computation Resources</a></li>
<li class="chapter" data-level="8.6" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>8.6</b> Extending Resources</a></li>
<li class="chapter" data-level="8.7" data-path="resourcer.html"><a href="resourcer.html#api-declaration-of-a-new-resource"><i class="fa fa-check"></i><b>8.7</b> API declaration of a new resource</a></li>
<li class="chapter" data-level="8.8" data-path="resourcer.html"><a href="resourcer.html#resource-forms"><i class="fa fa-check"></i><b>8.8</b> Resource Forms</a></li>
<li class="chapter" data-level="8.9" data-path="resourcer.html"><a href="resourcer.html#examples-with-different-resources"><i class="fa fa-check"></i><b>8.9</b> Examples with different resources</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>8.9.1</b> TSV file into a tibble or data.frame</a></li>
<li class="chapter" data-level="8.9.2" data-path="resourcer.html"><a href="resourcer.html#rdata-file-into-a-specific-r-object"><i class="fa fa-check"></i><b>8.9.2</b> .Rdata file into a specific R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Resources Extensions</b></span></li>
<li class="chapter" data-level="9" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>9</b> VCF files to GDS to peform GWAS with Bioconductor</a></li>
<li class="chapter" data-level="10" data-path="shell-extension.html"><a href="shell-extension.html"><i class="fa fa-check"></i><b>10</b> Secure shell programs: GWAS with PLINK</a></li>
<li class="part"><span><b>IV Workflows</b></span></li>
<li class="chapter" data-level="11" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>11</b> Setup</a>
<ul>
<li class="chapter" data-level="11.1" data-path="setup.html"><a href="setup.html#providing-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>11.1</b> Providing DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="11.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>11.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html"><i class="fa fa-check"></i><b>12</b> Basic statistical analyses</a>
<ul>
<li class="chapter" data-level="12.1" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-single-study"><i class="fa fa-check"></i><b>12.1</b> Analysis from a single study</a></li>
<li class="chapter" data-level="12.2" data-path="basic-statistical-analyses.html"><a href="basic-statistical-analyses.html#analysis-from-a-multiple-studies"><i class="fa fa-check"></i><b>12.2</b> Analysis from a multiple studies</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>13</b> Omic data analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="Omic.html"><a href="Omic.html#types-of-analyses-implemented"><i class="fa fa-check"></i><b>13.1</b> Types of analyses implemented</a></li>
<li class="chapter" data-level="13.2" data-path="Omic.html"><a href="Omic.html#differential-gene-expression-dge-analysis"><i class="fa fa-check"></i><b>13.2</b> Differential gene expression (DGE) analysis</a></li>
<li class="chapter" data-level="13.3" data-path="Omic.html"><a href="Omic.html#epigenome-wide-association-analysis-ewas"><i class="fa fa-check"></i><b>13.3</b> Epigenome-wide association analysis (EWAS)</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="Omic.html"><a href="Omic.html#single-cpg-analysis"><i class="fa fa-check"></i><b>13.3.1</b> Single CpG analysis</a></li>
<li class="chapter" data-level="13.3.2" data-path="Omic.html"><a href="Omic.html#multiple-cpg-analysis"><i class="fa fa-check"></i><b>13.3.2</b> Multiple CpG analysis</a></li>
<li class="chapter" data-level="13.3.3" data-path="Omic.html"><a href="Omic.html#adjusting-for-surrogate-variables"><i class="fa fa-check"></i><b>13.3.3</b> Adjusting for Surrogate Variables</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="Omic.html"><a href="Omic.html#gwas-with-bioconductor"><i class="fa fa-check"></i><b>13.4</b> GWAS with Bioconductor</a></li>
<li class="chapter" data-level="13.5" data-path="Omic.html"><a href="Omic.html#gwas-with-plink"><i class="fa fa-check"></i><b>13.5</b> GWAS with PLINK</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>14</b> Geographical data analysis</a></li>
<li class="part"><span><b>V Developers</b></span></li>
<li class="chapter" data-level="15" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>15</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="15.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>15.1</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="15.1.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>15.1.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="15.1.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>15.1.2</b> Test Datasets</a></li>
<li class="chapter" data-level="15.1.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>15.1.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="15.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>15.2</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>15.2.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="15.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>15.2.2</b> DataSHIELD Configuration</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>15.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="15.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>15.4</b> Debugging</a></li>
<li class="chapter" data-level="15.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>15.5</b> Limitations</a>
<ul>
<li class="chapter" data-level="15.5.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#function-parameters-parser"><i class="fa fa-check"></i><b>15.5.1</b> Function Parameters Parser</a></li>
<li class="chapter" data-level="15.5.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#server-side-environments"><i class="fa fa-check"></i><b>15.5.2</b> Server Side Environments</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>16</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="17" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html"><i class="fa fa-check"></i><b>17</b> Tips and tricks</a>
<ul>
<li class="chapter" data-level="17.1" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-install-r-packages-into-opal-server-from-r"><i class="fa fa-check"></i><b>17.1</b> How to install R packages into OPAL server from R</a></li>
<li class="chapter" data-level="17.2" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how-to-check-whether-there-are-open-r-sesions-in-the-opal-server"><i class="fa fa-check"></i><b>17.2</b> How to check whether there are open R sesions in the OPAL server</a></li>
<li class="chapter" data-level="17.3" data-path="tips-and-tricks.html"><a href="tips-and-tricks.html#how"><i class="fa fa-check"></i><b>17.3</b> How</a></li>
</ul></li>
<li class="part"><span><b>VI Appendix</b></span></li>
<li class="chapter" data-level="18" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>18</b> Session Info</a></li>
<li class="chapter" data-level="19" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>19</b> Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="basic-statistical-analyses" class="section level1" number="12">
<h1><span class="header-section-number">12</span> Basic statistical analyses</h1>
<p>Let us start by illustrating how to peform simple statistical data analyses using different resources. Here, we will use data from three studies that are avaialbe in our OPAL test repository. The three databases are called CNSIM1, CNSIM2, CNSIM3 that are avaialble as three different resources: mySQL database, SPSS file and CSV file (see Figure <a href="resources.html#fig:testResources">7.2</a>). This example mimics real situations were different hospital or research centers manage their own databases containing harmonized data. Data correspond to three simulated datasets with different number of observations of 11 harmonized variables. They contain synthetic data based on a model derived from the participants of the 1958 Birth Cohort, as part of the obesity methodological development project. This dataset does contain some NA values. The available variables are:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
Variable
</th>
<th style="text-align:left;">
Description
</th>
<th style="text-align:left;">
Type
</th>
<th style="text-align:left;">
Note
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LAB_TSC
</td>
<td style="text-align:left;">
Total Serum Cholesterol
</td>
<td style="text-align:left;">
numeric
</td>
<td style="text-align:left;">
mmol/L
</td>
</tr>
<tr>
<td style="text-align:left;">
LAB_TRIG
</td>
<td style="text-align:left;">
Triglycerides
</td>
<td style="text-align:left;">
numeric
</td>
<td style="text-align:left;">
mmol/L
</td>
</tr>
<tr>
<td style="text-align:left;">
LAB_HDL
</td>
<td style="text-align:left;">
HDL Cholesterol
</td>
<td style="text-align:left;">
numeric
</td>
<td style="text-align:left;">
mmol/L
</td>
</tr>
<tr>
<td style="text-align:left;">
LAB_GLUC_ADJUSTED
</td>
<td style="text-align:left;">
Non-Fasting Glucose
</td>
<td style="text-align:left;">
numeric
</td>
<td style="text-align:left;">
mmol/L
</td>
</tr>
<tr>
<td style="text-align:left;">
PM_BMI_CONTINUOUS
</td>
<td style="text-align:left;">
Body Mass Index (continuous)
</td>
<td style="text-align:left;">
numeric
</td>
<td style="text-align:left;">
kg/m2
</td>
</tr>
<tr>
<td style="text-align:left;">
DIS_CVA
</td>
<td style="text-align:left;">
History of Stroke
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
0 = Never had stroke; 1 = Has had stroke
</td>
</tr>
<tr>
<td style="text-align:left;">
MEDI_LPD
</td>
<td style="text-align:left;">
Current Use of Lipid Lowering Medication (from categorical assessment item)
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
0 = Not currently using lipid lowering medication; 1 = Currently using lipid lowering medication
</td>
</tr>
<tr>
<td style="text-align:left;">
DIS_DIAB
</td>
<td style="text-align:left;">
History of Diabetes
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
0 = Never had diabetes; 1 = Has had diabetes
</td>
</tr>
<tr>
<td style="text-align:left;">
DIS_AMI
</td>
<td style="text-align:left;">
History of Myocardial Infarction
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
0 = Never had myocardial infarction; 1 = Has had myocardial infarction
</td>
</tr>
<tr>
<td style="text-align:left;">
GENDER
</td>
<td style="text-align:left;">
Gender
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
0 = Female
</td>
</tr>
<tr>
<td style="text-align:left;">
PM_BMI_CATEGORICAL
</td>
<td style="text-align:left;">
Body Mass Index (categorical)
</td>
<td style="text-align:left;">
factor
</td>
<td style="text-align:left;">
1 = Less than 25 kg/m2; 2 = 25 to 30 kg/m2; 3 = Over 30 kg/m2
</td>
</tr>
</tbody>
</table>
<p>The analyses that are described here, can also be found in the <a href="https://data2knowledge.atlassian.net/wiki/spaces/DSDEV/pages/714571780/Tutorial+for+DataSHIELD+users+v5+-+http+bit.ly+intro-DS">DataSHIELD Tutorial</a> where these resources where uploaded into the Opal server as three tables, a much worse approach since data have to be moved from original repositories.</p>
<div id="analysis-from-a-single-study" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> Analysis from a single study</h2>
<p>Let us start by illustrating how to analyze one data set (CNSIM1).</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="basic-statistical-analyses.html#cb41-1"></a><span class="kw">library</span>(DSI)</span>
<span id="cb41-2"><a href="basic-statistical-analyses.html#cb41-2"></a><span class="kw">library</span>(DSOpal)</span>
<span id="cb41-3"><a href="basic-statistical-analyses.html#cb41-3"></a><span class="kw">library</span>(dsBaseClient)</span>
<span id="cb41-4"><a href="basic-statistical-analyses.html#cb41-4"></a></span>
<span id="cb41-5"><a href="basic-statistical-analyses.html#cb41-5"></a><span class="co"># prepare login data and resource to assign</span></span>
<span id="cb41-6"><a href="basic-statistical-analyses.html#cb41-6"></a>builder &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb41-7"><a href="basic-statistical-analyses.html#cb41-7"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, </span>
<span id="cb41-8"><a href="basic-statistical-analyses.html#cb41-8"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb41-9"><a href="basic-statistical-analyses.html#cb41-9"></a>               <span class="dt">resource =</span> <span class="st">&quot;test.CNSIM1&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb41-10"><a href="basic-statistical-analyses.html#cb41-10"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb41-11"><a href="basic-statistical-analyses.html#cb41-11"></a></span>
<span id="cb41-12"><a href="basic-statistical-analyses.html#cb41-12"></a><span class="co"># login and assign resource</span></span>
<span id="cb41-13"><a href="basic-statistical-analyses.html#cb41-13"></a>conns &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, </span>
<span id="cb41-14"><a href="basic-statistical-analyses.html#cb41-14"></a>                               <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span>
<span id="cb41-15"><a href="basic-statistical-analyses.html#cb41-15"></a></span>
<span id="cb41-16"><a href="basic-statistical-analyses.html#cb41-16"></a></span>
<span id="cb41-17"><a href="basic-statistical-analyses.html#cb41-17"></a><span class="co"># coerce ResourceClient objects to a data.frame called &#39;D&#39;</span></span>
<span id="cb41-18"><a href="basic-statistical-analyses.html#cb41-18"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;D&quot;</span>, </span>
<span id="cb41-19"><a href="basic-statistical-analyses.html#cb41-19"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.data.frame</span>(res)))</span></code></pre></div>
<p>Then we can inspect the type of data we have</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="basic-statistical-analyses.html#cb42-1"></a><span class="kw">ds.class</span>(<span class="st">&quot;D&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="basic-statistical-analyses.html#cb44-1"></a><span class="kw">ds.colnames</span>(<span class="st">&quot;D&quot;</span>)</span></code></pre></div>
<pre><code>$study1
 [1] &quot;entity_id&quot;          &quot;DIS_AMI&quot;            &quot;DIS_CVA&quot;            &quot;DIS_DIAB&quot;          
 [5] &quot;GENDER&quot;             &quot;LAB_GLUC_ADJUSTED&quot;  &quot;LAB_HDL&quot;            &quot;LAB_TRIG&quot;          
 [9] &quot;LAB_TSC&quot;            &quot;MEDI_LPD&quot;           &quot;PM_BMI_CATEGORICAL&quot; &quot;PM_BMI_CONTINUOUS&quot; </code></pre>
<p>Perform some data descriptive analyses</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="basic-statistical-analyses.html#cb46-1"></a><span class="kw">ds.table1D</span>(<span class="st">&quot;D$DIS_DIAB&quot;</span>)</span></code></pre></div>
<pre><code>$counts
      D$DIS_DIAB
0           3041
1             47
Total       3088

$percentages
      D$DIS_DIAB
0          98.48
1           1.52
Total     100.00

$validity
[1] &quot;All tables are valid!&quot;</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="basic-statistical-analyses.html#cb48-1"></a><span class="kw">ds.table2D</span>(<span class="st">&quot;D$DIS_DIAB&quot;</span>, <span class="st">&quot;D$GENDER&quot;</span>)</span></code></pre></div>
<pre><code>$colPercent
$colPercent$`study1-D$DIS_DIAB(row)|D$GENDER(col)`
           0      1  Total
0      98.04  98.94  98.48
1       1.96   1.06   1.52
Total 100.00 100.00 100.00


$colPercent.all.studies
$colPercent.all.studies$`pooled-D$DIS_DIAB(row)|D$GENDER(col)`
           0      1  Total
0      98.04  98.94  98.48
1       1.96   1.06   1.52
Total 100.00 100.00 100.00


$rowPercent
$rowPercent$`study1-D$DIS_DIAB(row)|D$GENDER(col)`
          0     1 Total
0     51.10 48.90   100
1     65.96 34.04   100
Total 51.33 48.67   100


$rowPercent.all.studies
$rowPercent.all.studies$`pooled-D$DIS_DIAB(row)|D$GENDER(col)`
          0     1 Total
0     51.10 48.90   100
1     65.96 34.04   100
Total 51.33 48.67   100


$chi2Test
$chi2Test$`study1-D$DIS_DIAB(row)|D$GENDER(col)`

    Pearson&#39;s Chi-squared test with Yates&#39; continuity correction

data:  contingencyTable
X-squared = 3.5158, df = 1, p-value = 0.06079



$chi2Test.all.studies
$chi2Test.all.studies$`pooled-D$DIS_DIAB(row)|D$GENDER(col)`

    Pearson&#39;s Chi-squared test with Yates&#39; continuity correction

data:  pooledContingencyTable
X-squared = 3.5158, df = 1, p-value = 0.06079



$counts
$counts$`study1-D$DIS_DIAB(row)|D$GENDER(col)`
         0    1 Total
0     1554 1487  3041
1       31   16    47
Total 1585 1503  3088


$counts.all.studies
$counts.all.studies$`pooled-D$DIS_DIAB(row)|D$GENDER(col)`
         0    1 Total
0     1554 1487  3041
1       31   16    47
Total 1585 1503  3088


$validity
[1] &quot;All tables are valid!&quot;</code></pre>
<p>Or even some statistical modelling. In this case we want to assess whether sex (GENDER) or trigrycerids (LAB_TRIG) are risk factors for diabetes (DIS_DIAB)</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="basic-statistical-analyses.html#cb50-1"></a>mod &lt;-<span class="st"> </span><span class="kw">ds.glm</span>(DIS_DIAB <span class="op">~</span><span class="st"> </span>LAB_TRIG <span class="op">+</span><span class="st"> </span>GENDER, </span>
<span id="cb50-2"><a href="basic-statistical-analyses.html#cb50-2"></a>       <span class="dt">data =</span> <span class="st">&quot;D&quot;</span> , <span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb50-3"><a href="basic-statistical-analyses.html#cb50-3"></a>mod<span class="op">$</span>coeff</span></code></pre></div>
<pre><code>              Estimate Std. Error    z-value      p-value low0.95CI.LP high0.95CI.LP        P_OR
(Intercept) -5.0254035 0.39487495 -12.726570 4.209299e-37   -5.7993442    -4.2514628 0.006526066
LAB_TRIG     0.3462886 0.09880155   3.504891 4.567950e-04    0.1526411     0.5399361 1.413810621
GENDER      -0.4550068 0.39497696  -1.151983 2.493280e-01   -1.2291474     0.3191338 0.634443636
            low0.95CI.P_OR high0.95CI.P_OR
(Intercept)     0.00302039      0.01404336
LAB_TRIG        1.16490687      1.71589723
GENDER          0.29254188      1.37593539</code></pre>
<p>As usual the connection must be closed</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="basic-statistical-analyses.html#cb52-1"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>
</div>
<div id="analysis-from-a-multiple-studies" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> Analysis from a multiple studies</h2>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="basic-statistical-analyses.html#cb53-1"></a><span class="kw">library</span>(DSOpal)</span>
<span id="cb53-2"><a href="basic-statistical-analyses.html#cb53-2"></a><span class="kw">library</span>(dsBaseClient)</span>
<span id="cb53-3"><a href="basic-statistical-analyses.html#cb53-3"></a></span>
<span id="cb53-4"><a href="basic-statistical-analyses.html#cb53-4"></a><span class="co"># prepare login data and resources to assign</span></span>
<span id="cb53-5"><a href="basic-statistical-analyses.html#cb53-5"></a>builder &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb53-6"><a href="basic-statistical-analyses.html#cb53-6"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, <span class="dt">resource =</span> <span class="st">&quot;test.CNSIM1&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb53-7"><a href="basic-statistical-analyses.html#cb53-7"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study2&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, <span class="dt">resource =</span> <span class="st">&quot;test.CNSIM2&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb53-8"><a href="basic-statistical-analyses.html#cb53-8"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study3&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, <span class="dt">resource =</span> <span class="st">&quot;test.CNSIM3&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb53-9"><a href="basic-statistical-analyses.html#cb53-9"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb53-10"><a href="basic-statistical-analyses.html#cb53-10"></a></span>
<span id="cb53-11"><a href="basic-statistical-analyses.html#cb53-11"></a><span class="co"># login and assign resources</span></span>
<span id="cb53-12"><a href="basic-statistical-analyses.html#cb53-12"></a>conns &lt;-<span class="st"> </span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span>
<span id="cb53-13"><a href="basic-statistical-analyses.html#cb53-13"></a></span>
<span id="cb53-14"><a href="basic-statistical-analyses.html#cb53-14"></a><span class="co"># assigned objects are of class ResourceClient (and others)</span></span>
<span id="cb53-15"><a href="basic-statistical-analyses.html#cb53-15"></a><span class="kw">ds.class</span>(<span class="st">&quot;res&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;SQLResourceClient&quot; &quot;ResourceClient&quot;    &quot;R6&quot;               

$study2
[1] &quot;TidyFileResourceClient&quot; &quot;FileResourceClient&quot;     &quot;ResourceClient&quot;        
[4] &quot;R6&quot;                    

$study3
[1] &quot;TidyFileResourceClient&quot; &quot;FileResourceClient&quot;     &quot;ResourceClient&quot;        
[4] &quot;R6&quot;                    </code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="basic-statistical-analyses.html#cb55-1"></a><span class="co"># coerce ResourceClient objects to data.frames</span></span>
<span id="cb55-2"><a href="basic-statistical-analyses.html#cb55-2"></a><span class="co"># (DataSHIELD config allows as.resource.data.frame() assignment function for the purpose of the demo)</span></span>
<span id="cb55-3"><a href="basic-statistical-analyses.html#cb55-3"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;D&quot;</span>, <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.data.frame</span>(res)))</span>
<span id="cb55-4"><a href="basic-statistical-analyses.html#cb55-4"></a><span class="kw">ds.class</span>(<span class="st">&quot;D&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;

$study2
[1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;

$study3
[1] &quot;spec_tbl_df&quot; &quot;tbl_df&quot;      &quot;tbl&quot;         &quot;data.frame&quot; </code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="basic-statistical-analyses.html#cb57-1"></a><span class="co"># note that some dsBase functions do not like that the data.frame has multiple and different classes</span></span>
<span id="cb57-2"><a href="basic-statistical-analyses.html#cb57-2"></a><span class="co"># (despite all are data.frames). Then query colnames one by one:</span></span>
<span id="cb57-3"><a href="basic-statistical-analyses.html#cb57-3"></a><span class="kw">lapply</span>(conns, <span class="cf">function</span>(conn) {<span class="kw">ds.colnames</span>(<span class="st">&quot;D&quot;</span>, <span class="dt">datasources =</span> conn)})</span></code></pre></div>
<pre><code>$study1
 [1] &quot;id&quot;                 &quot;LAB_TSC&quot;            &quot;LAB_TRIG&quot;           &quot;LAB_HDL&quot;           
 [5] &quot;LAB_GLUC_ADJUSTED&quot;  &quot;PM_BMI_CONTINUOUS&quot;  &quot;DIS_CVA&quot;            &quot;MEDI_LPD&quot;          
 [9] &quot;DIS_DIAB&quot;           &quot;DIS_AMI&quot;            &quot;GENDER&quot;             &quot;PM_BMI_CATEGORICAL&quot;

$study2
 [1] &quot;entity_id&quot;          &quot;DIS_AMI&quot;            &quot;DIS_CVA&quot;            &quot;DIS_DIAB&quot;          
 [5] &quot;GENDER&quot;             &quot;LAB_GLUC_ADJUSTED&quot;  &quot;LAB_HDL&quot;            &quot;LAB_TRIG&quot;          
 [9] &quot;LAB_TSC&quot;            &quot;MEDI_LPD&quot;           &quot;PM_BMI_CATEGORICAL&quot; &quot;PM_BMI_CONTINUOUS&quot; 

$study3
 [1] &quot;entity_id&quot;          &quot;DIS_AMI&quot;            &quot;DIS_CVA&quot;            &quot;DIS_DIAB&quot;          
 [5] &quot;GENDER&quot;             &quot;LAB_GLUC_ADJUSTED&quot;  &quot;LAB_HDL&quot;            &quot;LAB_TRIG&quot;          
 [9] &quot;LAB_TSC&quot;            &quot;MEDI_LPD&quot;           &quot;PM_BMI_CATEGORICAL&quot; &quot;PM_BMI_CONTINUOUS&quot; </code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="basic-statistical-analyses.html#cb59-1"></a><span class="co"># do usual dsBase analysis</span></span>
<span id="cb59-2"><a href="basic-statistical-analyses.html#cb59-2"></a><span class="kw">ds.summary</span>(<span class="st">&#39;D$LAB_HDL&#39;</span>)</span></code></pre></div>
<pre><code>$study1
$study1$class
[1] &quot;numeric&quot;

$study1$length
[1] 2163

$study1$`quantiles &amp; mean`
      5%      10%      25%      50%      75%      90%      95%     Mean 
0.875240 1.047400 1.300000 1.581000 1.844500 2.090000 2.210900 1.569416 


$study2
$study2$class
[1] &quot;numeric&quot;

$study2$length
[1] 3088

$study2$`quantiles &amp; mean`
      5%      10%      25%      50%      75%      90%      95%     Mean 
0.850280 1.032200 1.294000 1.563000 1.840000 2.077000 2.225000 1.556648 


$study3
$study3$class
[1] &quot;numeric&quot;

$study3$length
[1] 4128

$study3$`quantiles &amp; mean`
      5%      10%      25%      50%      75%      90%      95%     Mean 
0.876760 1.039200 1.304000 1.589000 1.856000 2.098800 2.244200 1.574687 </code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="basic-statistical-analyses.html#cb61-1"></a><span class="co"># vector types are not necessarily the same depending on the data reader that was used</span></span>
<span id="cb61-2"><a href="basic-statistical-analyses.html#cb61-2"></a><span class="kw">ds.class</span>(<span class="st">&#39;D$GENDER&#39;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;integer64&quot;

$study2
[1] &quot;haven_labelled&quot;

$study3
[1] &quot;numeric&quot;</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="basic-statistical-analyses.html#cb63-1"></a><span class="kw">ds.asFactor</span>(<span class="st">&#39;D$GENDER&#39;</span>, <span class="st">&#39;GENDER&#39;</span>)</span></code></pre></div>
<pre><code>$all.unique.levels
[1] &quot;0&quot; &quot;1&quot;

$return.message
[1] &quot;Data object &lt;GENDER&gt; correctly created in all specified data sources&quot;</code></pre>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="basic-statistical-analyses.html#cb65-1"></a><span class="kw">ds.summary</span>(<span class="st">&#39;GENDER&#39;</span>)</span></code></pre></div>
<pre><code>$study1
$study1$class
[1] &quot;factor&quot;

$study1$length
[1] 2163

$study1$categories
[1] &quot;0&quot; &quot;1&quot;

$study1$`count of &#39;0&#39;`
[1] 1092

$study1$`count of &#39;1&#39;`
[1] 1071


$study2
$study2$class
[1] &quot;factor&quot;

$study2$length
[1] 3088

$study2$categories
[1] &quot;0&quot; &quot;1&quot;

$study2$`count of &#39;0&#39;`
[1] 1585

$study2$`count of &#39;1&#39;`
[1] 1503


$study3
$study3$class
[1] &quot;factor&quot;

$study3$length
[1] 4128

$study3$categories
[1] &quot;0&quot; &quot;1&quot;

$study3$`count of &#39;0&#39;`
[1] 2091

$study3$`count of &#39;1&#39;`
[1] 2037</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="basic-statistical-analyses.html#cb67-1"></a><span class="co"># or coerce to a dplyr&#39;s tbl, which is more suitable for large/big datasets analysis</span></span>
<span id="cb67-2"><a href="basic-statistical-analyses.html#cb67-2"></a><span class="co"># (DataSHIELD config allows as.resource.tbl() assignment function for the purpose of the demo)</span></span>
<span id="cb67-3"><a href="basic-statistical-analyses.html#cb67-3"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;T&quot;</span>, <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.tbl</span>(res)))</span>
<span id="cb67-4"><a href="basic-statistical-analyses.html#cb67-4"></a><span class="kw">ds.class</span>(<span class="st">&quot;T&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;tbl_MariaDBConnection&quot; &quot;tbl_dbi&quot;               &quot;tbl_sql&quot;               &quot;tbl_lazy&quot;             
[5] &quot;tbl&quot;                  

$study2
[1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;

$study3
[1] &quot;spec_tbl_df&quot; &quot;tbl_df&quot;      &quot;tbl&quot;         &quot;data.frame&quot; </code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="basic-statistical-analyses.html#cb69-1"></a><span class="co"># DataSHIELD analysis using dplyr objects and functions is to be invented...</span></span>
<span id="cb69-2"><a href="basic-statistical-analyses.html#cb69-2"></a></span>
<span id="cb69-3"><a href="basic-statistical-analyses.html#cb69-3"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="setup.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="Omic.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
