<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 The resourcer package | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD</title>
  <meta name="description" content="7 The resourcer package | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="7 The resourcer package | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 The resourcer package | Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD" />
  
  
  



<meta name="date" content="2020-05-06" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="the-resources.html"/>
<link rel="next" href="setup.html"/>
<script src="libs/header-attrs-2.1/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Wellcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="rbioconductor-data-infrastructures.html"><a href="rbioconductor-data-infrastructures.html"><i class="fa fa-check"></i><b>2</b> R/Bioconductor data infrastructures</a></li>
<li class="chapter" data-level="3" data-path="opal.html"><a href="opal.html"><i class="fa fa-check"></i><b>3</b> OPAL</a></li>
<li class="chapter" data-level="4" data-path="datashield.html"><a href="datashield.html"><i class="fa fa-check"></i><b>4</b> DataSHIELD</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datashield.html"><a href="datashield.html#overview"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html"><i class="fa fa-check"></i><b>5</b> DSI: DataSHIELD Interface implementation for Opal data repository</a>
<ul>
<li class="chapter" data-level="5.1" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#class-structure"><i class="fa fa-check"></i><b>5.1</b> Class Structure</a></li>
<li class="chapter" data-level="5.2" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#higher-level-functions"><i class="fa fa-check"></i><b>5.2</b> Higher Level Functions</a></li>
<li class="chapter" data-level="5.3" data-path="dsi-datashield-interface-implementation-for-opal-data-repository.html"><a href="dsi-datashield-interface-implementation-for-opal-data-repository.html#options"><i class="fa fa-check"></i><b>5.3</b> Options</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="6" data-path="the-resources.html"><a href="the-resources.html"><i class="fa fa-check"></i><b>6</b> The resources</a>
<ul>
<li class="chapter" data-level="6.1" data-path="the-resources.html"><a href="the-resources.html#introduction-1"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="the-resources.html"><a href="the-resources.html#types-of-resources"><i class="fa fa-check"></i><b>6.2</b> Types of resources</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="resourcer.html"><a href="resourcer.html"><i class="fa fa-check"></i><b>7</b> The resourcer package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="resourcer.html"><a href="resourcer.html#file-resources"><i class="fa fa-check"></i><b>7.1</b> File Resources</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="resourcer.html"><a href="resourcer.html#file-getter"><i class="fa fa-check"></i><b>7.1.1</b> File Getter</a></li>
<li class="chapter" data-level="7.1.2" data-path="resourcer.html"><a href="resourcer.html#file-data-format"><i class="fa fa-check"></i><b>7.1.2</b> File Data Format</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="resourcer.html"><a href="resourcer.html#database-resources"><i class="fa fa-check"></i><b>7.2</b> Database Resources</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="resourcer.html"><a href="resourcer.html#dbi-connectors"><i class="fa fa-check"></i><b>7.2.1</b> DBI Connectors</a></li>
<li class="chapter" data-level="7.2.2" data-path="resourcer.html"><a href="resourcer.html#use-dplyr"><i class="fa fa-check"></i><b>7.2.2</b> Use dplyr</a></li>
<li class="chapter" data-level="7.2.3" data-path="resourcer.html"><a href="resourcer.html#document-databases"><i class="fa fa-check"></i><b>7.2.3</b> Document Databases</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="resourcer.html"><a href="resourcer.html#computation-resources"><i class="fa fa-check"></i><b>7.3</b> Computation Resources</a></li>
<li class="chapter" data-level="7.4" data-path="resourcer.html"><a href="resourcer.html#extending_resources"><i class="fa fa-check"></i><b>7.4</b> Extending Resources</a></li>
<li class="chapter" data-level="7.5" data-path="resourcer.html"><a href="resourcer.html#api-declaration-of-a-new-resource"><i class="fa fa-check"></i><b>7.5</b> API declaration of a new resource</a></li>
<li class="chapter" data-level="7.6" data-path="resourcer.html"><a href="resourcer.html#examples-with-different-resources"><i class="fa fa-check"></i><b>7.6</b> Examples with different resources</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="resourcer.html"><a href="resourcer.html#tsv-file-into-a-tibble-or-data.frame"><i class="fa fa-check"></i><b>7.6.1</b> TSV file into a tibble or data.frame</a></li>
<li class="chapter" data-level="7.6.2" data-path="resourcer.html"><a href="resourcer.html#rdata-file-into-a-specific-r-object"><i class="fa fa-check"></i><b>7.6.2</b> .Rdata file into a specific R object</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="8" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>8</b> Setup</a>
<ul>
<li class="chapter" data-level="8.1" data-path="setup.html"><a href="setup.html#required-datashield-packages-in-the-opal-server"><i class="fa fa-check"></i><b>8.1</b> Required DataSHIELD packages in the opal server</a></li>
<li class="chapter" data-level="8.2" data-path="setup.html"><a href="setup.html#required-r-packages-in-the-client-site-e.g.-local-machine"><i class="fa fa-check"></i><b>8.2</b> Required R Packages in the client site (e.g. local machine)</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="statistical-analyses-from-different-resources.html"><a href="statistical-analyses-from-different-resources.html"><i class="fa fa-check"></i><b>9</b> Statistical analyses from different resources</a></li>
<li class="chapter" data-level="10" data-path="Omic.html"><a href="Omic.html"><i class="fa fa-check"></i><b>10</b> Using the resources for transcriptomic and epigenomic data analysis with Bioconductor</a></li>
<li class="chapter" data-level="11" data-path="extending-the-resources-to-deal-with-vcf-file-in-genomic-data-analysis-gwas-with-bioconductor.html"><a href="extending-the-resources-to-deal-with-vcf-file-in-genomic-data-analysis-gwas-with-bioconductor.html"><i class="fa fa-check"></i><b>11</b> Extending the resources to deal with VCF file in genomic data analysis: GWAS with Bioconductor</a></li>
<li class="chapter" data-level="12" data-path="omic-extension.html"><a href="omic-extension.html"><i class="fa fa-check"></i><b>12</b> Extending the resources to SHELL programs: GWAS with PLINK</a></li>
<li class="chapter" data-level="13" data-path="GIS.html"><a href="GIS.html"><i class="fa fa-check"></i><b>13</b> Extending the resources to Geographical Information System (GIS) analyses</a></li>
<li class="part"><span><b>IV Developers</b></span></li>
<li class="chapter" data-level="14" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html"><i class="fa fa-check"></i><b>14</b> DSLite: DataSHIELD Implementation on Local Datasets</a>
<ul>
<li class="chapter" data-level="14.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#introduction-2"><i class="fa fa-check"></i><b>14.1</b> Introduction</a></li>
<li class="chapter" data-level="14.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#development-environment-setup"><i class="fa fa-check"></i><b>14.2</b> Development Environment Setup</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-packages"><i class="fa fa-check"></i><b>14.2.1</b> DataSHIELD Packages</a></li>
<li class="chapter" data-level="14.2.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#test-datasets"><i class="fa fa-check"></i><b>14.2.2</b> Test Datasets</a></li>
<li class="chapter" data-level="14.2.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#r-package-development-tools"><i class="fa fa-check"></i><b>14.2.3</b> R Package Development Tools</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-development-flow"><i class="fa fa-check"></i><b>14.3</b> DataSHIELD Development Flow</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#dsliteserver"><i class="fa fa-check"></i><b>14.3.1</b> DSLiteServer</a></li>
<li class="chapter" data-level="14.3.2" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-configuration"><i class="fa fa-check"></i><b>14.3.2</b> DataSHIELD Configuration</a></li>
<li class="chapter" data-level="14.3.3" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#datashield-sessions"><i class="fa fa-check"></i><b>14.3.3</b> DataSHIELD Sessions</a></li>
<li class="chapter" data-level="14.3.4" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#debugging"><i class="fa fa-check"></i><b>14.3.4</b> Debugging</a></li>
<li class="chapter" data-level="14.3.5" data-path="dslite-datashield-implementation-on-local-datasets.html"><a href="dslite-datashield-implementation-on-local-datasets.html#limitations"><i class="fa fa-check"></i><b>14.3.5</b> Limitations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="creating-datashield-packages.html"><a href="creating-datashield-packages.html"><i class="fa fa-check"></i><b>15</b> Creating DataSHIELD packages</a></li>
<li class="chapter" data-level="16" data-path="useful-information.html"><a href="useful-information.html"><i class="fa fa-check"></i><b>16</b> Useful information</a>
<ul>
<li class="chapter" data-level="16.1" data-path="useful-information.html"><a href="useful-information.html#how-to-install-r-packages-into-opal-server-from-r"><i class="fa fa-check"></i><b>16.1</b> How to install R packages into OPAL server from R</a></li>
<li class="chapter" data-level="16.2" data-path="useful-information.html"><a href="useful-information.html#how-to-check-whether-there-are-open-r-sesions-in-the-opal-server"><i class="fa fa-check"></i><b>16.2</b> How to check whether there are open R sesions in the OPAL server</a></li>
<li class="chapter" data-level="16.3" data-path="useful-information.html"><a href="useful-information.html#how"><i class="fa fa-check"></i><b>16.3</b> How</a></li>
</ul></li>
<li class="part"><span><b>V Appendix</b></span></li>
<li class="chapter" data-level="17" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>17</b> Session Info</a></li>
<li class="chapter" data-level="18" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>18</b> Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating non-disclosive big data analyses of shared data from different resources with R and DataSHIELD</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="resourcer" class="section level1" number="7">
<h1><span class="header-section-number">7</span> The resourcer package</h1>
<p>The <code>resourcer</code> package is meant to access resources identified by a URL in a uniform way whether it references a dataset (stored in a file, a SQL table, a MongoDB collection etc.) or a computation unit (system commands, web services etc.). Usually some credentials will be defined, and an additional data format information can be provided to help dataset coercing to a data.frame object.</p>
<p>The main concepts are:</p>
<ul>
<li><em>Resource</em>, access to a resource (dataset or computation unit) is described by an object with URL, optional credentials and optional data format properties,</li>
<li><em>ResourceResolver</em>, a <em>ResourceClient</em> factory based on the URL scheme and available in a resolvers registry,</li>
<li><em>ResourceClient</em>, realizes the connection with the dataset or the computation unit described by a <em>Resource</em>,</li>
<li><em>FileResourceGetter</em>, connect to a file described by a resource,</li>
<li><em>DBIResourceConnector</em>, establish a <a href="https://www.r-dbi.org/">DBI</a> connection.</li>
</ul>
<div id="file-resources" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> File Resources</h2>
<p>These are resources describing a file. If the file is in a remote location, it must be downloaded before being read. The data format specification of the resource helps to find the appropriate file reader.</p>
<div id="file-getter" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> File Getter</h3>
<p>The file locations supported by default are:</p>
<ul>
<li><code>file</code>, local file system,</li>
<li><code>http</code>(s), web address, basic authentication,</li>
<li><code>gridfs</code>, MongoDB file store,</li>
<li><code>scp</code>, file copy through SSH,</li>
<li><code>opal</code>, <a href="https://www.obiba.org/pages/products/opal/">Opal</a> file store.</li>
</ul>
<p>This can be easily applied to other file locations by extending the <em>FileResourceGetter</em> class. An instance of the new file resource getter is to be registered so that the <em>FileResourceResolver</em> can operate as expected.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="resourcer.html#cb1-1"></a>resourcer<span class="op">::</span><span class="kw">registerFileResourceGetter</span>(<span class="kw">MyFileLocationResourceGetter</span>()<span class="op">$</span><span class="kw">new</span>())</span></code></pre></div>
</div>
<div id="file-data-format" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> File Data Format</h3>
<p>The data format specified within the <em>Resource</em> object, helps at finding the appropriate file reader. Currently supported data formats are:</p>
<ul>
<li>the data formats that have a reader in <a href="https://www.tidyverse.org/">tidyverse</a>: <a href="https://readr.tidyverse.org/">readr</a> (<code>csv</code>, <code>csv2</code>, <code>tsv</code>, <code>ssv</code>, <code>delim</code>), <a href="https://haven.tidyverse.org/">haven</a> (<code>spss</code>, <code>sav</code>, <code>por</code>, <code>dta</code>, <code>stata</code>, <code>sas</code>, <code>xpt</code>), <a href="https://readxl.tidyverse.org/">readxl</a> (<code>excel</code>, <code>xls</code>, <code>xlsx</code>). This can be easily applied to other data file formats by extending the <em>FileResourceClient</em> class.</li>
<li>the R data format that can be loaded in a child R environment from which object of interest will be retrieved.</li>
</ul>
<p>Usage example that reads a local SPSS file:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="resourcer.html#cb2-1"></a><span class="co"># make a SPSS file resource</span></span>
<span id="cb2-2"><a href="resourcer.html#cb2-2"></a>res &lt;-<span class="st"> </span>resourcer<span class="op">::</span><span class="kw">newResource</span>(</span>
<span id="cb2-3"><a href="resourcer.html#cb2-3"></a>  <span class="dt">name =</span> <span class="st">&quot;CNSIM1&quot;</span>,</span>
<span id="cb2-4"><a href="resourcer.html#cb2-4"></a>  <span class="dt">url =</span> <span class="st">&quot;file:///data/CNSIM1.sav&quot;</span>,</span>
<span id="cb2-5"><a href="resourcer.html#cb2-5"></a>  <span class="dt">format =</span> <span class="st">&quot;spss&quot;</span></span>
<span id="cb2-6"><a href="resourcer.html#cb2-6"></a>)</span>
<span id="cb2-7"><a href="resourcer.html#cb2-7"></a><span class="co"># coerce the csv file in the opal server to a data.frame</span></span>
<span id="cb2-8"><a href="resourcer.html#cb2-8"></a>df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(res)</span></code></pre></div>
<p>To support other file data format, extend the <em>FileResourceClient</em> class with the new data format reader implementation. Associate factory class, an extension of the <em>ResourceResolver</em> class is also to be implemented and registered.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="resourcer.html#cb3-1"></a>resourcer<span class="op">::</span><span class="kw">registerResourceResolver</span>(MyFileFormatResourceResolver<span class="op">$</span><span class="kw">new</span>())</span></code></pre></div>
</div>
</div>
<div id="database-resources" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Database Resources</h2>
<div id="dbi-connectors" class="section level3" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> DBI Connectors</h3>
<p><a href="https://www.r-dbi.org/">DBI</a> is a set of virtual classes that are are used to abstract the SQL database connections and operations within R. Then any DBI implementation can be used to access to a SQL table. Which DBI connector to be used is an information that can be extracted from the scheme part of the resource’s URL. For instance a resource URL starting with <code>postgres://</code> will require the <a href="https://rpostgres.r-dbi.org/">RPostgres</a> driver. To separate the DBI connector instanciation from the DBI interface interactions in the <em>SQLResourceClient</em>, a <em>DBIResourceConnector</em> registry is to be populated. The currently supported SQL database connectors are:</p>
<ul>
<li><code>mariadb</code> MariaDB connector,</li>
<li><code>mysql</code> MySQL connector,</li>
<li><code>postgres</code> or <code>postgresql</code> Postgres connector,</li>
<li><code>presto</code>, <code>presto+http</code> or <code>presto+https</code> <a href="https://prestodb.io/">Presto</a> connector,</li>
<li><code>spark</code>, <code>spark+http</code> or <code>spark+https</code> <a href="https://spark.apache.org/">Spark</a> connector.</li>
</ul>
<p>To support another SQL database having a DBI driver, extend the <em>DBIResourceConnector</em> class and register it:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="resourcer.html#cb4-1"></a>resourcer<span class="op">::</span><span class="kw">registerDBIResourceConnector</span>(MyDBResourceConnector<span class="op">$</span><span class="kw">new</span>())</span></code></pre></div>
</div>
<div id="use-dplyr" class="section level3" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Use dplyr</h3>
<p>Having the data stored in the database allows to handle large (common SQL databases) to big (PrestoDB, Spark) datasets using <a href="https://dplyr.tidyverse.org/">dplyr</a> which will delegate as much as possible operations to the database.</p>
</div>
<div id="document-databases" class="section level3" number="7.2.3">
<h3><span class="header-section-number">7.2.3</span> Document Databases</h3>
<p>NoSQL databases can be described by a resource. The <a href="https://docs.ropensci.org/nodbi/">nodbi</a> can be used here. Currently only connection to MongoDB database is supported using URL scheme <code>mongodb</code> or <code>mongodb+srv</code>.</p>
</div>
</div>
<div id="computation-resources" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Computation Resources</h2>
<p>Computation resources are resources on which tasks/commands can be triggerred and from which resulting data can be retrieved.</p>
<p>Example of computation resource that connects to a server through SSH:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="resourcer.html#cb5-1"></a><span class="co"># make an application resource on a ssh server</span></span>
<span id="cb5-2"><a href="resourcer.html#cb5-2"></a>res &lt;-<span class="st"> </span>resourcer<span class="op">::</span><span class="kw">newResource</span>(</span>
<span id="cb5-3"><a href="resourcer.html#cb5-3"></a>  <span class="dt">name =</span> <span class="st">&quot;supercomp1&quot;</span>,</span>
<span id="cb5-4"><a href="resourcer.html#cb5-4"></a>  <span class="dt">url =</span> <span class="st">&quot;ssh://server1.example.org/work/dir?exec=plink,ls&quot;</span>,</span>
<span id="cb5-5"><a href="resourcer.html#cb5-5"></a>  <span class="dt">identity =</span> <span class="st">&quot;sshaccountid&quot;</span>,</span>
<span id="cb5-6"><a href="resourcer.html#cb5-6"></a>  <span class="dt">secret =</span> <span class="st">&quot;sshaccountpwd&quot;</span></span>
<span id="cb5-7"><a href="resourcer.html#cb5-7"></a>)</span>
<span id="cb5-8"><a href="resourcer.html#cb5-8"></a><span class="co"># get ssh client from resource object</span></span>
<span id="cb5-9"><a href="resourcer.html#cb5-9"></a>client &lt;-<span class="st"> </span>resourcer<span class="op">::</span><span class="kw">newResourceClient</span>(res) <span class="co"># does a ssh::ssh_connect()</span></span>
<span id="cb5-10"><a href="resourcer.html#cb5-10"></a><span class="co"># execute commands</span></span>
<span id="cb5-11"><a href="resourcer.html#cb5-11"></a>files &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">exec</span>(<span class="st">&quot;ls&quot;</span>) <span class="co"># exec &#39;cd /work/dir &amp;&amp; ls&#39;</span></span>
<span id="cb5-12"><a href="resourcer.html#cb5-12"></a><span class="co"># release connection</span></span>
<span id="cb5-13"><a href="resourcer.html#cb5-13"></a>client<span class="op">$</span><span class="kw">close</span>() <span class="co"># does ssh::ssh_disconnect(session)</span></span></code></pre></div>
</div>
<div id="extending_resources" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Extending Resources</h2>
<p>There are several ways to extend the Resources handling. These are based on different R6 classes having a <code>isFor(resource)</code> function:</p>
<ul>
<li>If the resource is a file located at a place not already handled, write a new <em>FileResourceGetter</em> subclass and register an instance of it with the function <code>registerFileResourceGetter()</code>.</li>
<li>If the resource is a SQL engine having a DBI connector defined, write a new <em>DBIResourceConnector</em> subclass and register an instance of it with the function <code>registerDBIResourceConnector()</code>.</li>
<li>If the resource is in a domain specific web application or database, write a new <em>ResourceResolver</em> subclass and register an instance of it with the function <code>registerResourceResolver()</code>. This <em>ResourceResolver</em> object will create the appropriate <em>ResourceClient</em> object that matches your needs.</li>
</ul>
<p>The design of the URL that will describe your new resource should not overlap an existing one, otherwise the different registries will return the first instance for which the <code>isFor(resource)</code> is <code>TRUE</code>. In order to distinguish resource locations, the URL’s scheme can be extended, for instance the scheme for accessing a file in a Opal server is <code>opal+https</code> so that the credentials be applied as needed by Opal.</p>
</div>
<div id="api-declaration-of-a-new-resource" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> API declaration of a new resource</h2>
<p>It is possible to declare a resource that is to be resolved by an R package that uses the <code>resourcer</code> API. This figure illustrate how to declare a new reource called VCF2GDS that load Variant Calling Format files into Genomic Data Storage objects that can be managed whiting R. This will be further illustrated when analyzing omic data analysis in this <a href="omic-extension.html#omic_extension">Section</a>.</p>
<div class="figure" style="text-align: center"><span id="fig:testDeclaration"></span>
<img src="fig/opal_resources_API.png" alt="Declaration of a resource corresponding to a VCF2GDS format" width="680" />
<p class="caption">
Figure 7.1: Declaration of a resource corresponding to a VCF2GDS format
</p>
</div>
</div>
<div id="examples-with-different-resources" class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Examples with different resources</h2>
<p>Let us illustrate how to deal with different types of resources within DataSHIELD. To this end, let use our Opal test example available at <a href="https://opal-test.obiba.org" class="uri">https://opal-test.obiba.org</a> which has the following reources</p>
<div class="figure" style="text-align: center"><span id="fig:testResources2"></span>
<img src="fig/opal_resources.png" alt="Resources from a test enviroment available at https://opal-test.obiba.org"  />
<p class="caption">
Figure 7.2: Resources from a test enviroment available at <a href="https://opal-test.obiba.org" class="uri">https://opal-test.obiba.org</a>
</p>
</div>
<p>Let us start by illustrating how to get a simple TSV file (brge.txt) into the R server. This file is located at a GitHub repository: <a href="https://raw.githubusercontent.com/isglobal-brge/brgedata/master/inst/extdata/brge.txt" class="uri">https://raw.githubusercontent.com/isglobal-brge/brgedata/master/inst/extdata/brge.txt</a> and it is not necesary to be moved from there. This is one of the main strenght of the resources implementation.</p>
<div id="tsv-file-into-a-tibble-or-data.frame" class="section level3" number="7.6.1">
<h3><span class="header-section-number">7.6.1</span> TSV file into a tibble or data.frame</h3>
<p>This code describes how to get the resource (a TSV file) as a data.frame into the R Server. Note that this is a secure access since user name and password must be provided</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="resourcer.html#cb6-1"></a><span class="kw">library</span>(DSI)</span>
<span id="cb6-2"><a href="resourcer.html#cb6-2"></a><span class="kw">library</span>(DSOpal)</span>
<span id="cb6-3"><a href="resourcer.html#cb6-3"></a><span class="kw">library</span>(dsBaseClient)</span>
<span id="cb6-4"><a href="resourcer.html#cb6-4"></a></span>
<span id="cb6-5"><a href="resourcer.html#cb6-5"></a><span class="co"># access to the &#39;brge&#39; resource (</span><span class="al">NOTE</span><span class="co">: test.brge is need since the project</span></span>
<span id="cb6-6"><a href="resourcer.html#cb6-6"></a><span class="co"># is called test)</span></span>
<span id="cb6-7"><a href="resourcer.html#cb6-7"></a>builder &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb6-8"><a href="resourcer.html#cb6-8"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, </span>
<span id="cb6-9"><a href="resourcer.html#cb6-9"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb6-10"><a href="resourcer.html#cb6-10"></a>               <span class="dt">resource =</span> <span class="st">&quot;test.brge&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb6-11"><a href="resourcer.html#cb6-11"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb6-12"><a href="resourcer.html#cb6-12"></a></span>
<span id="cb6-13"><a href="resourcer.html#cb6-13"></a><span class="co"># the resource is loaded into R as the object &#39;res&#39; </span></span>
<span id="cb6-14"><a href="resourcer.html#cb6-14"></a>conns &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, </span>
<span id="cb6-15"><a href="resourcer.html#cb6-15"></a>                               <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span>
<span id="cb6-16"><a href="resourcer.html#cb6-16"></a></span>
<span id="cb6-17"><a href="resourcer.html#cb6-17"></a><span class="co"># the resource is assigned to a data.frame</span></span>
<span id="cb6-18"><a href="resourcer.html#cb6-18"></a><span class="co"># Assign to the original R class (e.g ExpressionSet)</span></span>
<span id="cb6-19"><a href="resourcer.html#cb6-19"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;dat&quot;</span>, </span>
<span id="cb6-20"><a href="resourcer.html#cb6-20"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.data.frame</span>(res)))</span>
<span id="cb6-21"><a href="resourcer.html#cb6-21"></a></span>
<span id="cb6-22"><a href="resourcer.html#cb6-22"></a><span class="kw">ds.class</span>(<span class="st">&quot;dat&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;spec_tbl_df&quot; &quot;tbl_df&quot;      &quot;tbl&quot;         &quot;data.frame&quot; </code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="resourcer.html#cb8-1"></a><span class="co"># logout the connection</span></span>
<span id="cb8-2"><a href="resourcer.html#cb8-2"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>
</div>
<div id="rdata-file-into-a-specific-r-object" class="section level3" number="7.6.2">
<h3><span class="header-section-number">7.6.2</span> .Rdata file into a specific R object</h3>
<p>Now let us describe how to get an specific type of R object into de R server. Our Opal test contains a resource called GSE80970 which is in a local machine. The resource is an R object of class <a href="https://kasperdanielhansen.github.io/genbioconductor/html/ExpressionSet.html">ExpressionSet</a> which is normally used to jointly capsulate gene expression, metadata and annotation. In general, we can retrieve any R object in their original format and if a method to coerce the specific object into a data.frame exists, we can also retrieve it as a tibble/data.frame.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="resourcer.html#cb9-1"></a><span class="co"># prepare login data and resource to assign</span></span>
<span id="cb9-2"><a href="resourcer.html#cb9-2"></a>builder &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">newDSLoginBuilder</span>()</span>
<span id="cb9-3"><a href="resourcer.html#cb9-3"></a>builder<span class="op">$</span><span class="kw">append</span>(<span class="dt">server =</span> <span class="st">&quot;study1&quot;</span>, <span class="dt">url =</span> <span class="st">&quot;https://opal-test.obiba.org&quot;</span>, </span>
<span id="cb9-4"><a href="resourcer.html#cb9-4"></a>               <span class="dt">user =</span> <span class="st">&quot;dsuser&quot;</span>, <span class="dt">password =</span> <span class="st">&quot;password&quot;</span>, </span>
<span id="cb9-5"><a href="resourcer.html#cb9-5"></a>               <span class="dt">resource =</span> <span class="st">&quot;test.GSE80970&quot;</span>, <span class="dt">driver =</span> <span class="st">&quot;OpalDriver&quot;</span>)</span>
<span id="cb9-6"><a href="resourcer.html#cb9-6"></a>logindata &lt;-<span class="st"> </span>builder<span class="op">$</span><span class="kw">build</span>()</span>
<span id="cb9-7"><a href="resourcer.html#cb9-7"></a></span>
<span id="cb9-8"><a href="resourcer.html#cb9-8"></a><span class="co"># login and assign resource (to &#39;res&#39; symbol)</span></span>
<span id="cb9-9"><a href="resourcer.html#cb9-9"></a>conns &lt;-<span class="st"> </span>DSI<span class="op">::</span><span class="kw">datashield.login</span>(<span class="dt">logins =</span> logindata, <span class="dt">assign =</span> <span class="ot">TRUE</span>, </span>
<span id="cb9-10"><a href="resourcer.html#cb9-10"></a>                               <span class="dt">symbol =</span> <span class="st">&quot;res&quot;</span>)</span>
<span id="cb9-11"><a href="resourcer.html#cb9-11"></a></span>
<span id="cb9-12"><a href="resourcer.html#cb9-12"></a><span class="co"># coerce ResourceClient objects to a data.frame called &#39;DF&#39; </span></span>
<span id="cb9-13"><a href="resourcer.html#cb9-13"></a><span class="co"># </span><span class="al">NOTE</span><span class="co">: as.data.frame exists for `ExpressionSet` objects</span></span>
<span id="cb9-14"><a href="resourcer.html#cb9-14"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;DF&quot;</span>,  </span>
<span id="cb9-15"><a href="resourcer.html#cb9-15"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.data.frame</span>(res)))</span>
<span id="cb9-16"><a href="resourcer.html#cb9-16"></a></span>
<span id="cb9-17"><a href="resourcer.html#cb9-17"></a><span class="kw">ds.class</span>(<span class="st">&quot;DF&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="resourcer.html#cb11-1"></a><span class="co"># we can also coerce ResourceClient objects to their original format.</span></span>
<span id="cb11-2"><a href="resourcer.html#cb11-2"></a><span class="co"># This will allow the analyses with specific R/Bioconductor packages</span></span>
<span id="cb11-3"><a href="resourcer.html#cb11-3"></a><span class="kw">datashield.assign.expr</span>(conns, <span class="dt">symbol =</span> <span class="st">&quot;ES&quot;</span>, </span>
<span id="cb11-4"><a href="resourcer.html#cb11-4"></a>                       <span class="dt">expr =</span> <span class="kw">quote</span>(<span class="kw">as.resource.object</span>(res)))</span>
<span id="cb11-5"><a href="resourcer.html#cb11-5"></a></span>
<span id="cb11-6"><a href="resourcer.html#cb11-6"></a><span class="kw">ds.class</span>(<span class="st">&quot;ES&quot;</span>)</span></code></pre></div>
<pre><code>$study1
[1] &quot;ExpressionSet&quot;
attr(,&quot;package&quot;)
[1] &quot;Biobase&quot;</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="resourcer.html#cb13-1"></a><span class="co"># logout the connection</span></span>
<span id="cb13-2"><a href="resourcer.html#cb13-2"></a><span class="kw">datashield.logout</span>(conns)</span></code></pre></div>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="the-resources.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="setup.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
